stumpy/scrump.py

```diff
@@ -472,46 +568,90 @@ def scrump(T, m, percentage=0.01, pre_scrump=False, s=None):
     See Algorithm 1 and Algorithm 2
     """
 
-    T = np.asarray(T)
-    T = T.copy()
-    T[np.isinf(T)] = np.nan
-    core.check_dtype(T)
+    if T_B is not None and pre_scrump:  # pragma: no cover
+        raise NotImplementedError(
+            "AB joins with preprocessing are not yet implemented."
+        )
+
+    T_A = np.asarray(T_A)
+    if T_A.ndim != 1:  # pragma: no cover
+        raise ValueError(
+            f"T_A is {T_A.ndim}-dimensional and must be 1-dimensional. "
+            "For multidimensional STUMP use `stumpy.mstump` or `stumpy.mstumped`"
+        )
 
-    if T.ndim != 1:  # pragma: no cover
-        raise ValueError(f"T is {T.ndim}-dimensional and must be 1-dimensional. ")
+    T_A = T_A.copy()
+    T_A[np.isinf(T_A)] = np.nan
+    core.check_dtype(T_A)
+
+    if T_B is None:
+        T_B = T_A
+        ignore_trivial = True
+
+    T_B = np.asarray(T_B)
+    T_B = T_B.copy()
+    if T_B.ndim != 1:  # pragma: no cover
+        raise ValueError(
+            f"T_B is {T_B.ndim}-dimensional and must be 1-dimensional. "
+            "For multidimensional STUMP use `stumpy.mstump` or `stumpy.mstumped`"
+        )
+    T_B[np.isinf(T_B)] = np.nan
+    core.check_dtype(T_B)
 
     core.check_window_size(m)
 
-    n = T.shape[0]
-    l = n - m + 1
+    if ignore_trivial is False and core.are_arrays_equal(T_A, T_B):  # pragma: no cover
+        logger.warning("Arrays T_A, T_B are equal, which implies a self-join.")
+        logger.warning("Try setting `ignore_trivial = True`.")
+
+    if ignore_trivial and core.are_arrays_equal(T_A, T_B) is False:  # pragma: no cover
+        logger.warning("Arrays T_A, T_B are not equal, which implies an AB-join.")
+        logger.warning("Try setting `ignore_trivial = False`.")
+
+    n_A = T_A.shape[0]
+    n_B = T_B.shape[0]
+    l = n_B - m + 1
+
     out = np.empty((l, 2), dtype=object)
     out[:, 0] = np.inf
     out[:, 1] = -1
 
-    M_T, Σ_T = core.compute_mean_std(T, m)
-    T[np.isnan(T)] = 0
+    M_T, Σ_T = core.compute_mean_std(T_A, m)
+    μ_Q, σ_Q = core.compute_mean_std(T_B, m)
+
+    T_A[np.isnan(T_A)] = 0
+    T_B[np.isnan(T_B)] = 0
+
     excl_zone = int(np.ceil(m / 4))
 
     if s is None:
         s = excl_zone
 
     if pre_scrump:
-        P, I = prescrump(T, m, M_T, Σ_T, s)
+        P, I = prescrump(T_A, m, M_T, Σ_T, s)
         for i in range(P.shape[0]):
             if out[i, 0] > P[i]:
                 out[i, 0] = P[i]
                 out[i, 1] = I[i]
 
-    orders = np.random.permutation(range(excl_zone + 1, n - m + 1))
+    if ignore_trivial:
+        orders = np.random.permutation(range(excl_zone + 1, n_B - m + 1))
+    else:
+        orders = np.random.permutation(range(-(n_A - m + 1) + 1, n_B - m + 1))
```

Comment:
Do you mind explaining the logic of having negative values in `orders`? Each element of `orders` should be referring to a diagonal. So, I am assuming that a negative value refers to the lower triangle. However, typically, NumPy indexing with negative numbers refers to indexing from the end of an array. In this case, IIUC, it looks like `orders` is relative to the main diagonal that runs through `[0, 0]` and has nothing to do with NumPy indexing? Is that about right?

Reply:
That's correct. Take a look at the following distance matrix:

```
 0   1   2   3   4
-1   0   1   2   3
-2  -1   0   1   2
```
Then the length of the Matrix profile is 5 (number of columns = `n_B - m +1`) while the number of rows is `n_A - m + 1`.

Reply:
Cool. Thanks